tfhe/src/core_crypto/fft_impl/f128/math/fft/tests.rs

```diff
@@ -0,0 +1,213 @@
+
+            let mut fourier0_re0 = avec![0.0f64; fourier_size].into_boxed_slice();
+            let mut fourier0_re1 = avec![0.0f64; fourier_size].into_boxed_slice();
+            let mut fourier0_im0 = avec![0.0f64; fourier_size].into_boxed_slice();
+            let mut fourier0_im1 = avec![0.0f64; fourier_size].into_boxed_slice();
+
+            let mut fourier1_re0 = avec![0.0f64; fourier_size].into_boxed_slice();
+            let mut fourier1_re1 = avec![0.0f64; fourier_size].into_boxed_slice();
+            let mut fourier1_im0 = avec![0.0f64; fourier_size].into_boxed_slice();
+            let mut fourier1_im1 = avec![0.0f64; fourier_size].into_boxed_slice();
+
+            for (x, y) in izip!(poly0.as_mut().iter_mut(), poly1.as_mut().iter_mut()) {
+                *x = generator.random_uniform();
+                *y = generator.random_uniform();
+
+                *y >>= Scalar::BITS - 16;
+            }
+
+            let mut mem = GlobalPodBuffer::new(fft.backward_scratch().unwrap());
+            let mut stack = PodStack::new(&mut mem);
+
+            fft.forward_as_torus(
+                &mut fourier0_re0,
+                &mut fourier0_re1,
+                &mut fourier0_im0,
+                &mut fourier0_im1,
+                &poly0,
+            );
+            fft.forward_as_integer(
+                &mut fourier1_re0,
+                &mut fourier1_re1,
+                &mut fourier1_im0,
+                &mut fourier1_im1,
+                &poly1,
+            );
+
+            for (f0_re0, f0_re1, f0_im0, f0_im1, f1_re0, f1_re1, f1_im0, f1_im1) in izip!(
+                &mut *fourier0_re0,
+                &mut *fourier0_re1,
+                &mut *fourier0_im0,
+                &mut *fourier0_im1,
+                &*fourier1_re0,
+                &*fourier1_re1,
+                &*fourier1_im0,
+                &*fourier1_im1,
+            ) {
+                let f0_re = f128(*f0_re0, *f0_re1);
+                let f0_im = f128(*f0_im0, *f0_im1);
+                let f1_re = f128(*f1_re0, *f1_re1);
+                let f1_im = f128(*f1_im0, *f1_im1);
+
+                f128(*f0_re0, *f0_re1) = f0_re * f1_re - f0_im * f1_im;
+                f128(*f0_im0, *f0_im1) = f0_im * f1_re + f0_re * f1_im;
+            }
+
+            fft.backward_as_torus(
+                &mut convolution_from_fft,
+                &fourier0_re0,
+                &fourier0_re1,
+                &fourier0_im0,
+                &fourier0_im1,
+                stack.rb_mut(),
+            );
+            convolution_naive(
+                convolution_from_naive.as_mut(),
+                poly0.as_ref(),
+                poly1.as_ref(),
+            );
+
+            for (expected, actual) in izip!(
+                convolution_from_naive.as_ref().iter(),
+                convolution_from_fft.as_ref().iter()
+            ) {
+                let threshold = Scalar::ONE << (Scalar::BITS - 5);
```

Comment:
I'm still a bit confused by this one, how many correct bits do we expect to recover ? About a 100 right ? so the error should probably more something like 1 << (Scalar::BITS - 100) ?

Reply:
well that formula stands only for u128, but if we expect to recover the exact results for u32 and u64 then we should test it this way, and have the appropriate thershold for u128

Reply:
we lose a few bits from the multiplication by an integer of magnitude ~ 2^16, so i think a good threshold would be
`Scalar::ONE << (Scalar::BITS.saturating_sub(100 - integer_magnitude - 4))`

(tests pass with this value)

Reply:
then fine by me !

I don't want to have a flaky tests but it felt like we were losing quite a lot from the formula, I let you be the judge for the security margin adding a comment explaining the reasoning for the threshold 🙂

Reply:
this seems like a more reasonable formula
```
                let threshold = Scalar::ONE
                    << (Scalar::BITS.saturating_sub(100 - integer_magnitude - (size_log + 1) / 2));
```
`size_log` is the log of the polynomial size, `(size_log + 1) / 2` is the error we expect to come from the fft

Reply:
perfect!